### DISCUSSION TITLE
skywalking-swck-controller-manager unavailability leading to ReplicaFailure

### DISCUSSION THREAD

**Discussion by saowu** [apache/skywalking#12001] [Q&A] [answered]
skywalking-swck-controller-manager is unavailable, all services in the namespace are running normally, but it is not possible to trigger the behavior of starting a new pod (such as restarting, scale Deployment, etc.)
```
Error creating: Internal error occurred: failed calling webhook "mpod.kb.io": Post "https://skywalking-swck-webhook-service.skywalking-swck-system.svc:443/mutate-v1-pod?timeout=10s": no endpoints available for service "skywalking-swck-webhook-service"
```
Is it reasonable for pod startup to rely on skywalking swck controller manager？In theory, the service side should not rely on SWCK.

**Comment by wu-sheng** [comment]
No, we don't have codes to limit that inside oap pods, AFAIK.

**Reply by saowu** [reply]
When skywalk-swck-controller-manager is unavailable, Pods cannot be started in namespace with the swck-injection=enabled Labels added.

![image](https://github.com/apache/skywalking/assets/33569619/ea88ec95-5fe5-42f7-b61d-a9c709b90858)
![image](https://github.com/apache/skywalking/assets/33569619/ba1d40d2-1b33-4dcc-be5f-a6fcc3fe7a0d)

**Reply by wu-sheng** [reply]
If that is(I don't know), it may be affected by operator mechanism, not OAP side.

**Reply by saowu** [reply]
Yes, it's not OAP, it's skywalking-swck-controller-manager.

**Reply by saowu** [reply]
Is there a way to optimize it? It feels a bit dangerous.

**Reply by wu-sheng** [reply]
I don't know swck part, sorry.

**Reply by Superskyyy** [reply]
@dashanji any suggestions?

**Comment by dashanji** [comment, marked as answer]
Hi, @saowu. It's an expected behavior, as the webhook is contained in the `skywalking-swck-controller-manager`. If you don't want to use the manager or injector anymore( as you set the replicas of controller manager as 0), just delete the label of the namespace refer to https://github.com/apache/skywalking-swck/blob/ed08076b74334e7d18383c3e32fe617ffc7b8015/docs/java-agent-injector.md#active-the-java-agent-injection

**Reply by saowu** [reply]
That's right, but the availability of skywalking-swck-controller-management will affect the services on the probe side, creating dependencies. Is it safe.

**Reply by dashanji** [reply]
> That's right, but the availability of skywalking-swck-controller-management will affect the services on the probe side, creating dependencies. Is it safe.

The services and controller-manager are a whole, you can't imagine they can work independently. If you find the controller can't work, it's better to fix it.

**Reply by saowu** [reply]
Okay, I understand. Thank you

**Reply by saowu** [reply]
Excuse me, one more question. What are the image version list of SwAgent? The default version on the official website is 8.16.0-Java8. Is there an lastest version available? Can I build the image myself? I haven't found any relevant documents.

**Reply by dashanji** [reply]
Refer the doc https://github.com/apache/skywalking-swck/blob/master/docs/java-agent-injector.md#configure-sidecar

**Reply by saowu** [reply]
Sorry! I couldn't find what I wanted. Is there any other version of this image available? What is the version number? I want to build this image myself, what should I do.
![image](https://github.com/apache/skywalking/assets/33569619/19b297ac-11b7-43d3-9ce1-f94e011f5279)

**Reply by wu-sheng** [reply]
All images are from https://hub.docker.com/r/apache/skywalking-java-agent

**Reply by saowu** [reply]
Thank you, this is the answer I want to get! Thank you very much for your help！

**Reply by wu-sheng** [reply]
You could follow the CI file to control the rebuild, https://github.com/apache/skywalking-java/blob/main/.github/workflows/publish-docker.yaml. Makefile is available in the root.

**Reply by saowu** [reply]
Surprisingly, it's here. I've been searching for it in skywalking-swck, but I couldn't find it

**Reply by wu-sheng** [reply]
Are you asking for Java agent image build script? If so, yes. Java agent repository hosts that.

**Comment by saowu** [comment]
Yes, thank you


吴延博
***@***.***




------------------ Original ------------------
From: 吴晟 Wu Sheng ***@***.***&gt;
Date: Mon,Mar 18,2024 6:03 PM
To: apache/skywalking ***@***.***&gt;
Cc: 吴延博 ***@***.***&gt;, Mention ***@***.***&gt;
Subject: Re: [apache/skywalking] skywalking-swck-controller-managerunavailability leading to ReplicaFailure (Discussion #12001)





 
Are you asking for Java agent image build script? If so, yes. Java agent repository hosts that.
